src/main/java/com/uid2/admin/vertx/service/EncryptionKeyService.java

```diff
@@ -316,22 +332,24 @@ private RotationResult rotateKeys(Predicate<Integer> siteSelector, Duration minA
             return result;
         }
 
-        result.rotatedKeys = addSiteKeys(siteIds, activatesIn, expiresAfter);
+        result.rotatedKeys = addSiteKeys(siteIds, activatesIn, expiresAfter, cutoffTime);
 
         return result;
     }
 
-    private List<EncryptionKey> addSiteKeys(Iterable<Integer> siteIds, Duration activatesIn, Duration expiresAfter)
+    private List<EncryptionKey> addSiteKeys(Iterable<Integer> siteIds, Duration activatesIn, Duration expiresAfter, Duration cutoffTime)
             throws Exception {
+        final Instant now = clock.now();
+
         final List<EncryptionKey> keys = this.keyProvider.getSnapshot().getActiveKeySet().stream()
                 .sorted(Comparator.comparingInt(EncryptionKey::getId))
+                .filter(filterKeyOverCutOffTime ? k -> now.compareTo(k.getExpires().plus(cutoffTime.toDays(), ChronoUnit.DAYS)) < 0 : k -> true)
```

Comment:
This is where the feature flag/switch comes in. We will only filter keys out when we set `filterKeyOverCutOffTime` to be true.

Reply:
I think it would be better to replace the lambda with a method reference so we can have a descriptive name here. It takes quite a bit of brainwork to figure out what it actually does.
Probably also better to just have the function return true in a guard clause than have a ternary returning an always true lambda.

Reply:
To think of it, probably worth extracting the entire filtering functionality into a separate collaborator class/method. This file is already having way too much in it, it wouldn't make sense for your PR to fix all of that of course, but perhaps we can stop adding to the problem. Then we could also add unit tests rather than rely solely on rather lengthy and slow integration tests for everything. Happy to explain what I mean if you need help.